File: Recursion/007.Permutation/StringPermutation.swift
//
//  Permutation
//
//  Created by Feialoh Francis on 23/11/22.
//

import Foundation

class Permutation {
    
    var str: String
    
    init(str: String) {
        self.str = str
    }
    
    func permuations (_ pos: Int = 0, _ ans: [String] = [String]()) -> [String] {
        var ans = ans
        var str = str
        if pos >= str.count {
            ans.append(str)
        }
        
        for i in pos..<str.count {
            str.swapAt(i, pos)
            self.str = str
            ans = permuations(pos+1, ans)
            str.swapAt(i, pos)
        }
        return ans
    }
}

var allString = [String]()
print("Enter the string:")
let str = readLine()

if let str = str {
    print("Possible permutations are: ")
    let sol = Permutation(str: str)
    allString = sol.permuations()
    print(allString)
}


extension String {
    mutating func swapAt(_ index1: Int, _ index2: Int) {
        var characters = Array(self)
        characters.swapAt(index1, index2)
        self = String(characters)
    }
}

File: Recursion/007.Permutation/StringPermutation.cpp
#include<iostream>
#include<vector>
#include<bits/stdc++.h>
using namespace std;

void helper(int pos, string &str, vector<string> &ans){
    
    if (pos >= str.size()){
        ans.push_back(str);
    }
    unordered_set <char> m;
    for (int i=pos; i<str.size(); i++){
        if (m.find(str[i]) != m.end()) continue;
        m.insert(str[i]);
        swap(str[i],str[pos]);
        helper(pos+1, str, ans);
        swap(str[i],str[pos]);
    }
}

vector<string> permutation(string &str){
    vector<string> ans;
    helper(0, str, ans);
    sort(ans.begin(),ans.end());
    return ans;
}

int main() {

    string str;
    vector<string> allString;
    cout<<"Enter the string: ";
    cin>>str;
    cout<<"Possible permutations are: \n";
    allString = permutation(str);
    for(string s: allString){
        cout<<s<<"\n";
    }
    cout<<"Total = "<<allString.size();
    return 0;
}